### PR TITLE
Fix 16281 - std.stdio.readf should use ref instead of requiring pointers

### PR DESCRIPTION
**Pull request by Bolpat**
Making the pointer-taking versions of `formattedRead` and `readf` deprecated. That's what the bug is actually saying.

Keeping two versions that behave equivalent is inconsistent and there is no benefit of using pointers.

### CODE REVIEW

File: std/format.d [hunk 1]
```diff
@@ -569,66 +569,70 @@ matching failure happens.
 
 Throws:
     An `Exception` if `S.length == 0` and `fmt` has format specifiers
+
+Deprecated:
+
+The deprecation targets parameters of the form `&param` which take the address of
+a local and make most usages of this function non-@safe. Replace these parameters
```
Comment (wilzbach) [inline std/format.d]: s/a local/a parameter/

File: std/format.d [hunk 2]
```diff
@@ -569,66 +569,70 @@ matching failure happens.
...
+simply by `param`.
```
Comment (wilzbach) [inline std/format.d]: /simply/d (we try to use formal language for the official docs)

File: std/format.d [hunk 3]
```diff
@@ -569,66 +569,70 @@ matching failure happens.
...
+Paramters of the form `paramFunc()` are false positives; replace them by
+`*paramFunc()` to surpass the deprecation warning correctly.
```
Comment (wilzbach) [inline std/format.d]: Hmm this could be quite difficult to justify. Have you tried to add this to the if constraints?

File: std/format.d (hunk 1 above)
Comment (Bolpat) [inline std/format.d]: The deprecation (noun sg.) targets (verb in 3rd person) parameters (noun pl.) ...
I haven't had problems until now, but you're right. That's hard to read. I'll reformulate this.

File: std/format.d (hunk 2 above)
Comment (Bolpat) [inline std/format.d]: Done.

File: std/format.d (hunk 3 above)
Comment (Bolpat) [inline std/format.d]: This is impossible to distinguish from the viewpoint of the function.

**Comment by Bolpat** [comment]
From the viewpoint of the function, a parameter can be an l-value or an r-value and it can have pointer-type or not. So we have four cases to consider.

| | **Pointer** | **Non-Pointer** |
|-:|:-|:-|
| **L-Value** | *Unsupported* possible future use case. | Supposed to be the use case. |
| **R-Value** | Comes typically from `&var`, but could also be `returnsPtr()`. | *Compile-time error:* Cannot write to r-value. No overload takes those parameters. |

The form `returnsPtr()` of a parameter is *not* a false positive as mentioned in the deprecation description. Consider the following:
1. Assume `returnsPtr()` returns a pointer *by value.* Then we get the deprecation message. Reasons for the usage are:
    1. Old code using the old interface to read the content of the returned pointer. This is intensionally deprecated and should be replaced by `*returnsPtr()`.
    2. New code using the new interface to read the pointer itself. This is an error because the thing to be read must be an l-value.

This is only a false positive in the sense that the error in ii. is mistaken for the deprecated old usage. I don't see a problem at all because any programmer must admit that not all conceptual errors are recognized by the compiler. In this case, the it will even give a hint that something is wrong.

2. Assume `returnsPtr()` returns a pointer *by reference.* This is also the case for literal pointer variables. We have a compile-time error here.
The old interface has read into the content of the pointer as in 1. because it doesn't care about the origin of the pointer.
The new interface would try to read the pointer itself as it is an l-value.

Because, fortunately, `unformatValue` doesn't support pointers, a compile-time error occurs. Breakage is impossible: only heavily broken code that better: shouldn't have compiled would now compile with (probably) another meaning than intended.
If we don't allow reading pointers via `unformatValue` until this deprecation is an error, there will be no breakage or change of behavior.

**Comment by Bolpat** [comment]
For the "condition necessary" push, see [this forum question](https://forum.dlang.org/post/mcelzqirufuktwwhlkfz@forum.dlang.org).
I thought, `T*` was an obviously better match than `ref T` and the `auto ref` plus constraint can be omitted.

**Comment by andralex** [comment]
Let's not deprecate the pointer versions. They can coexist with the versions taking `ref`, and we can marginalize them in the documentation later.

What we need to do here is disable formatted reading of pointers for the `ref`-taking version. It has no business in most code (if you really want just read a hex integral and cast on your side) and makes format unsafe. I think we can put up with a breakage here - cc @MartinNowak.

So that means if you ever pass a pointer to `format`, be it an lvalue or rvalue, it goes to the (current) version of format. There's no ambiguity and it can be done inside `format` with `static if`.

**Comment by andralex** [comment]
@Bolpat on second thought how does one read pointers with readf? I just tried this and it doesn't compile:

```d
void main()
{
    import std.stdio;
    int *p;
    readf("%p", &p);
}
```

That would mean there's no ability to read pointers as integers, which means (a) there's no ambiguity between the `ref` and the pointer versions of `readf`, and (b) we can make even the pointer version of `readf` trusted. cc @byebye

**Comment by Bolpat** [comment]
@andralex:
> Let's not deprecate the pointer versions.

Why? That's progress. We can keep it deprecated as long as we whish. Not deprecating something means (at least to me) encouraging to use it. But taking the address of a local should not be encouraged; the cases where the pointers are not of the form `&x` are probably very rare. This interface is good for C, but it's weird for languages with references like D.

I can accept forwarding l-value pointers to the old interface. This usage should compile with a deprecation message.

From your answer in the discussion of [issue 8471](https://issues.dlang.org/show_bug.cgi?id=8471), I can accept making it trusted. Like @jmdavis, I'm skeptical towards giving templates the `@trusted` attribute.
The access to `stdin` can (should) be trusted as you say, but that's it. That should make `readf` safe. If it doesn't, the problem is located elsewhere, probably in `unformatValue`. Looking at the [spec](https://dlang.org/library/std/format/unformat_value.html), one cannot read user defined types (not even enums), but we maybe (hopefully) add this feature in the future (some `fromString` function?). `readf` for user defined types obviously cannot be trusted in general, suggesting not to trust it generically because actually `readf` doesn't care about what it specifically reads.

**Comment by andralex** [comment]
>> Let's not deprecate the pointer versions.

>Why? That's progress.

That debate has closed a while ago.

**Comment by Bolpat** [comment]
@andralex I'm sorry I haven't seen the debate or its result. I've just been searching the forum and PRs without result. Do you have a link for me? Like that, it's hard to accept for me.

**Comment by andralex** [comment]
@Bolpat thanks for your consideration. There's not one single link as much as a years-long evolving disposition. The short of it is this APi stays and we have the opportunity to add the new API.

The old API is even sound technically - taking the address of a local and passing it to a trusted function is an entirely fine thing to go about things because we know readf does not escape the addresses. If we defined such an API right now we'd use `scope` to clarify that. Thanks!

**Comment by andralex** [comment]
@Bolpat sorry this doesn't float your boat. @RazvanN7 can you please take this over? Thanks!

**Comment by Bolpat** [comment]
I could reopen as well. I interpreted your answer as a refusal. My reaction was precipitous.

**Comment by RazvanN7** [comment]
@Bolpat you need to do a rebase[1] since you have some conflicts.
@andralex which branch should we continue on?

[1] https://wiki.dlang.org/Starting_as_a_Contributor#Rebasing

**Comment by Bolpat** [comment]
@RazvanN7 I've done the merge wrong. Is rebasing really necessary?
Apart from that, I'd like to continue this pull. I can get credit for work like this at university.

**Comment by RazvanN7** [comment]
What exactly went wrong? It seems that the rebasing was successful since you no longer have any conflicts with the master branch.

**Comment by Bolpat** [comment]
I used the GitHub merge tool. Doesn't matter, does it? I had some bad experiences with rebase.

**Comment by Bolpat** [comment]
@andralex's answer was unclear to me. I interpreted it to reopen and discuss the (dis)advantages of deprecating the old interface.
Closing.

**Pull request closed without merging**